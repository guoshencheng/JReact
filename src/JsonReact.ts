import * as React from 'react';

export type JsonReactEventHandler = (key: string, data: any) => void;

export type JsonReactEvent = {
  handler: JsonReactEventHandler
} 

export type JsonReactComponent = {
  Cls: React.ComponentType<any>
}

export type JsonReact = {
  components: {
    [key: string]: JsonReactComponent,
  },
  events: {
    [key: string]: JsonReactEvent,
  }
}