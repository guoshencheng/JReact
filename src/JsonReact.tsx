import * as React from 'react';

export type ComponentJson = {
  data?: any,
  type: string,
  props?: PropsJson,
  events?: EventJson[],
  children?: ComponentJson | ComponentJson []
}

export type EventJson = {
  key: string,
  type: string,
}

export type PropsJson = {
  [key: string]: any,
}

export type StringMap<T> = {
  [key: string]: T
}

export type JsonReactEventHandler = (key: string, data: any) => void;

export type JsonReactEvent = {
  originEventKey: string,
  handler: JsonReactEventHandler
} 

export type JsonReactComponent = {
  Cls: React.ComponentType<any>
}

export class JsonReact {

  components: StringMap<JsonReactComponent>
  events: StringMap<JsonReactEvent>

  constructor() {
    this.components = {};
    this.events = {};
  }

  createJsonArrayComp(json: ComponentJson[]): React.ReactElement<any>[] {
    return json.map((item, index) => (
      this.createSingleJsonComp(item, index)
    ));
  }

  createSingleJsonComp(json: ComponentJson, key?: any): React.ReactElement<any> {
    const { components, events } = this;
    const { type, props, events: es, children, data } = json;
    const childNode = this.createJsonComp(children as any);
    const eventProps = es ? es.filter(e => !!events[e.type]).reduce((pre, cur) => {
      const ev = events[cur.type];
      pre[ev.originEventKey] = ev.handler;
      return pre;
    }, {} as { [key: string]: JsonReactEventHandler }) : {};
    const component = components[type];
    let Cls;
    if (!component) {
      Cls = type as keyof React.ReactHTML;
    } else {
      Cls = component.Cls;
    }
    if (typeof Cls === 'string') {
      return React.createElement(Cls, { ...props, ...eventProps, key }, childNode);
    } else {
      return (
        childNode ? <Cls data={data} key={key} {...props} {...eventProps}>{childNode}</Cls> : <Cls data={data} key={key} {...props} {...eventProps} />
      )
    }
  }

  createJsonComp(json: ComponentJson[]): React.ReactElement<any>[];
  createJsonComp(json: ComponentJson): React.ReactElement<any>;
  createJsonComp(): undefined;
  createJsonComp(json?: ComponentJson | ComponentJson[]): React.ReactElement<any>[] | React.ReactElement<any> | undefined {
    if (!json) return;
    if (!Array.isArray(json)) {
      return this.createSingleJsonComp(json);
    }
    return this.createJsonArrayComp(json);
  }

  component = (key: string, jsonReactComp: JsonReactComponent) => {
    this.components[key] = jsonReactComp;
  }

  event = (key: string, event: JsonReactEvent) => {
    this.events[key] = event;
  }
}
