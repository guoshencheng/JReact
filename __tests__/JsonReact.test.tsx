import JsonReact from '../src';
import * as ReactDOMServer from 'react-dom/server';
import * as trenderer from 'react-test-renderer';

import * as A from '../examples/components/A';
import * as B from '../examples/components/B';

JsonReact.RegisterComponent(A);
JsonReact.RegisterComponent(B);

describe('success render with json', () => {
  it('success render null as null', () => {
    const renderer = new JsonReact();
    const el = renderer.render();
    expect(
      ReactDOMServer.renderToString(el)
    ).toEqual('');
  });

  it('render a array type of json', () => {
    const renderer = new JsonReact();
    const el = renderer.render([{
      type: 'div',
      props: { style: { left: 20 } }
    }, {
      type: 'div',
      props: { style: { left: 50 } }
    }])
    const tree = trenderer.create(el).toJSON();
    expect(tree).toMatchSnapshot();
  })
  it('render div success', () => {
    const renderer = new JsonReact();
    const el = renderer.render({
      type: 'div',
      props: { style: { left: 20 } }
    })
    const tree = trenderer.create(el).toJSON();
    expect(tree).toMatchSnapshot();
  }) 
})

describe('render custom component', () => {
  const renderer = new JsonReact();
  const el = renderer.render([{
    type: 'div',
    props: { style: { left: 20 } },
    children: [
      'WelCome ',
      'To ',
      'JsonReact'
    ]
  }, {
    type: 'A',
    props: { style: { color: 'white', background: 'red' } },
    events: [{
      event: A.eventKeys.ACLICK.key,
      reducer: B.actionKeys.BCLICKED.key,
    }]
  }, {
    type: 'B',
  }]);
  it('success render custom component', () => {
    const tree = trenderer.create(el).toJSON();
    expect(tree).toMatchSnapshot();
  })
  it('change rendere result when dispatch event', () => {
    renderer.store.dispatch({
      type: B.actionKeys.BCLICKED.key
    })
    const tree = trenderer.create(el).toJSON();
    expect(tree).toMatchSnapshot();
  })
})