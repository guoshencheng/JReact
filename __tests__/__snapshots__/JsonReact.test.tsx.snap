// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render custom component change rendere result when dispatch event 1`] = `
Array [
  <div
    style={
      Object {
        "left": 20,
      }
    }
  >
    WelCome 
    To 
    JsonReact
  </div>,
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "red",
        "color": "white",
      }
    }
  >
    Click me!
  </div>,
  <div>
    clicked 
    1
  </div>,
]
`;

exports[`render custom component success render custom component 1`] = `
Array [
  <div
    style={
      Object {
        "left": 20,
      }
    }
  >
    WelCome 
    To 
    JsonReact
  </div>,
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "red",
        "color": "white",
      }
    }
  >
    Click me!
  </div>,
  <div>
    clicked 
    0
  </div>,
]
`;

exports[`success render with json render a array type of json 1`] = `
Array [
  <div
    style={
      Object {
        "left": 20,
      }
    }
  />,
  <div
    style={
      Object {
        "left": 50,
      }
    }
  />,
]
`;

exports[`success render with json render div success 1`] = `
<div
  style={
    Object {
      "left": 20,
    }
  }
/>
`;
